lib/Dialect/Seq/SeqOps.cpp

```diff
@@ -717,6 +717,15 @@ OpFoldResult ClockGateOp::fold(FoldAdaptor adaptor) {
   if (isConstantZero(adaptor.getInput()))
     return IntegerAttr::get(IntegerType::get(getContext(), 1), 0);
 
+  // Transitive clock gating - eliminate clock gates that are driven by an
+  // identical enable signal somewhere in a clock gate hierarchy.
+  auto clockGateInput = getInput().getDefiningOp<ClockGateOp>();
+  while (clockGateInput) {
+    if (clockGateInput.getEnable() == getEnable())
+      return getInput();
+    clockGateInput = clockGateInput.getInput().getDefiningOp<ClockGateOp>();
+  }
```

Comment:
I think you also need to check the `test_enable` here.

Reply:
hmm... would it then be correct to OR the `test_enable` of the current clock gate with the TE of the upstream matching clock gate?

Reply:
Yeah I think you could also do that! Or just check if the test enables also agree, not just the regular enable. I'd expect in most designs there is no test enable, or if there is one, it's all connected to one single signal almost everywhere in the hierarchy. So you might not be losing much if you just check for it instead of creating a new comb.or 😃